File: Domains/Base/Workspace/CSharp/Domain/Base/Workspace/WorkspaceObject.cs
namespace Allors.Workspace
{
    using System.Collections.Generic;
    using Allors.Meta;
    using Allors.Protocol.Remote.Sync;

    public class WorkspaceObject {
        public Workspace Workspace { get; }

        public long Id { get; }

        public long Version { get; }

        public string UserSecurityHash { get; }

        public Class ObjectType { get; }

        public Dictionary<string, object> Roles { get; }

        public Dictionary<string, object> Methods { get; }

        public WorkspaceObject(Workspace workspace, SyncResponse loadResponse, SyncResponseObject loadObject) {
            this.Workspace = workspace;
            this.Id = long.Parse(loadObject.I);
            this.Version = !string.IsNullOrEmpty(loadObject.V) ? long.Parse(loadObject.V) : 0;
            this.UserSecurityHash = loadResponse.UserSecurityHash;
            this.ObjectType = (Class)this.Workspace.ObjectFactory.GetObjectTypeForTypeName(loadObject.T);

            this.Roles = new Dictionary<string, object>();
            this.Methods =new Dictionary<string, object>();

            if (loadObject.Roles != null)
            {
                foreach (var role in loadObject.Roles)
                {
                    var name = (string)role[0];
                    var access = (string)role[1];
                    var canRead = access.Contains("r");
                    var canWrite = access.Contains("w");

                    this.Roles[$"CanRead{name}"] = canRead;
                    this.Roles[$"CanWrite{name}"] = canWrite;

                    if (canRead)
                    {
                        var value = role.Length > 2 ? role[2] : null;
                        this.Roles[name] = value;
                    }
                }
            }

            if (loadObject.Methods != null)
            {
                foreach (var method in loadObject.Methods)
                {
                    var name = method[0];
                    var access = method[1];
                    var canExecute = access.Contains("x");

                    this.Methods[$"CanExecute{name}"] = canExecute;
                }
            }
        }

        public bool CanRead(string roleTypeName) {
            return (bool)this.Roles[$"CanRead{roleTypeName}"];
        }

        public bool CanWrite(string roleTypeName) {
            return (bool)this.Roles[$"CanWrite{roleTypeName}"];
        }

        public bool CanExecute(string methodTypeName)
        {
            return (bool)this.Methods[$"CanExecute{methodTypeName}"];
        }
    }
}

File: Platform/Workspace/Typescript/src/allors/framework/workspace/WorkspaceObject.ts
import { ObjectType } from '../meta';
import { SyncResponse, SyncResponseObject } from '../protocol/sync/SyncResponse';
import { IWorkspace } from './Workspace';

export interface IWorkspaceObject {
    id: string;
    version: string;
    userSecurityHash: string;
    objectType: ObjectType;

    workspace: IWorkspace;
    roles: any;
    methods: any;

    canRead(roleTypeName: string): boolean;
    canWrite(roleTypeName: string): boolean;
    canExecute(methodName: string): boolean;
}

export class WorkspaceObject implements IWorkspaceObject {
    public workspace: IWorkspace;
    public roles: any;
    public methods: any;

    private i: string;
    private v: string;
    private u: string;
    private t: string;

    constructor(workspace: IWorkspace, loadResponse: SyncResponse, loadObject: SyncResponseObject) {
        this.workspace = workspace;
        this.i = loadObject.i;
        this.v = loadObject.v;
        this.u = loadResponse.userSecurityHash;
        this.t = loadObject.t;

        this.roles = {};
        this.methods = {};

        const objectType = this.workspace.metaPopulation.objectTypeByName[this.t];

        if (loadObject.roles) {
            loadObject.roles.forEach((role) => {
                const [name, access] = role;
                const canRead = access.indexOf('r') !== -1;
                const canWrite = access.indexOf('w') !== -1;

                this.roles[`CanRead${name}`] = canRead;
                this.roles[`CanWrite${name}`] = canWrite;

                if (canRead) {
                    const value = role[2];
                    this.roles[name] = value;
                }

            });
        }

        if (loadObject.methods) {
            loadObject.methods.forEach((method) => {
                const [name, access] = method;
                const canExecute = access.indexOf('x') !== -1;

                this.methods[`CanExecute${name}`] = canExecute;
            });
        }
    }

    get id(): string {
        return this.i;
    }

    get version(): string {
        return this.v;
    }

    get userSecurityHash(): string {
        return this.u;
    }

    get objectType(): ObjectType {
        return this.workspace.metaPopulation.objectTypeByName[this.t];
    }

    public canRead(roleTypeName: string): boolean {
        return this.roles[`CanRead${roleTypeName}`];
    }

    public canWrite(roleTypeName: string): boolean {
        return this.roles[`CanWrite${roleTypeName}`];
    }

    public canExecute(methodName: string): boolean {
        return this.methods[`CanExecute${methodName}`];
    }
}
